### PR TITLE
BAMloader

### PR DESCRIPTION
**Pull request by KitVB**
## Added BAM Loader

Added BAM Loaders and unit tests

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/data/data_loader.py [hunk 1]
```diff
@@ -1858,3 +1866,129 @@ def _featurize_shard(self, shard):
         else:
             x = DFTEntry.create(e_type, true_val, systems)
         return [x]
+
+
+class SAMLoader(DataLoader):
+    """Handles loading of SAM files.
+    
+    This class provides methods to load and featurize data from a SAM file.
```
Comment (rbharath) [inline deepchem/data/data_loader.py:1874]: Let's add more information to the documentation here. Add a usage example

File: deepchem/data/data_loader.py [hunk 2]
```diff
@@ -1858,3 +1866,129 @@ def _featurize_shard(self, shard):
...
+                # (X, y, w, ids)
+                yield X, None, None, ids
+
+        return DiskDataset.create_dataset(shard_generator(), data_dir)
+
+
+class BAMLoader(DataLoader):
+    """Handles loading of BAM files.
+    
+    This class provides methods to load and featurize data from a BAM file.
```
Comment (rbharath) [inline deepchem/data/data_loader.py:1937]: Same here (more docs and usage example)

File: deepchem/data/data_loader.py [hunk 3]
```diff
@@ -1858,3 +1866,129 @@ def _featurize_shard(self, shard):
         else:
             x = DFTEntry.create(e_type, true_val, systems)
         return [x]
+
+
+class SAMLoader(DataLoader):
```
Comment (rbharath) [inline deepchem/data/data_loader.py:1871]: This class needs to be added to the rst docs in the docs/ folder

File: deepchem/data/data_loader.py [hunk 4]
```diff
@@ -1858,3 +1866,129 @@ def _featurize_shard(self, shard):
...
+                samfile = pysam.AlignmentFile(input_file, "r")
+                X = self.featurizer.get_features(samfile)
+                ids = np.ones(len(X))
+                # (X, y, w, ids)
+                yield X, None, None, ids
+
+        return DiskDataset.create_dataset(shard_generator(), data_dir)
+
+
+class BAMLoader(DataLoader):
```
Comment (rbharath) [inline deepchem/data/data_loader.py:1934]: This class needs to be added to the rst docs in the docs/ folder

File: deepchem/feat/bio_seq_featurizer.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
+import numpy as np
+import deepchem as dc
+try:
+    import pysam
+except ImportError:
+    print("Error: Unable to import pysam. Please make sure it is installed.")
+
+
+class SAMFeaturizer:
```
Comment (rbharath) [inline deepchem/feat/bio_seq_featurizer.py:9]: This class needs to be added to the rst docs in the docs/ folder

File: deepchem/feat/bio_seq_featurizer.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    """
+    A class for extracting features from a SAM file.
```
Comment (rbharath) [inline deepchem/feat/bio_seq_featurizer.py:11]: Add more details here; what is it extracting? Add a usage example

File: deepchem/feat/bio_seq_featurizer.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+                break
+
+        samfile.close()
+
+        return np.array(features, dtype="object")
+
+
+class BAMFeaturizer:
+    """
+    A class for extracting features from a BAM file.
```
Comment (rbharath) [inline deepchem/feat/bio_seq_featurizer.py:101]: Same here (docs/usage example/ adding to rst docs)

File: requirements/env_common.yml [hunk 8]
```diff
@@ -27,3 +27,4 @@ dependencies:
     - xgboost
     - gensim  # used by mol2vec
     - tensorboard
+    - pysam
```
Comment (rbharath) [inline requirements/env_common.yml:30]: pysam needs to be added to the docs as a dependency. Is this a soft dependency?

File: deepchem/feat/bio_seq_featurizer.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+            # Break the loop if max_records is set
+            if self.max_records is not None and record_count >= self.max_records:
+                break
+
+        samfile.close()
+
+        return np.array(features, dtype="object")
+
+
+class BAMFeaturizer:
```
Comment (rbharath) [inline deepchem/feat/bio_seq_featurizer.py:99]: This doesn't inherit from Featurizer the abstract superclass

File: deepchem/feat/bio_seq_featurizer.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+        Initialize BAMFeaturizer.
+
+        Parameters
+        ----------
+        max_records : int or None, optional
+            The maximum number of records to extract from the BAM file. If None, all records will be extracted.
+        """
+        self.max_records = max_records
+
+    def get_features(self, bamfile):
```
Comment (rbharath) [inline deepchem/feat/bio_seq_featurizer.py:148]: Don't have your own API; use DeepChem's existing featurizer API

File: deepchem/feat/bio_seq_featurizer.py (hunk 5 above)
Comment (rbharath) [inline deepchem/feat/bio_seq_featurizer.py:9]: Same here

File: deepchem/data/__init__.py [hunk 11]
```diff
@@ -21,3 +21,5 @@
 from deepchem.data.data_loader import FASTALoader
 from deepchem.data.data_loader import ImageLoader
 from deepchem.data.data_loader import InMemoryLoader
+from deepchem.data.data_loader import SAMLoader
+from deepchem.data.data_loader import BAMLoader
```
Comment (shreyasvinaya) [inline deepchem/data/__init__.py:25]: need to add this in a try catch, might break imports if pysam is not installed

File: deepchem/feat/__init__.py [hunk 12]
```diff
@@ -85,3 +85,6 @@
 
 # support classes
 from deepchem.feat.molecule_featurizers import GraphMatrix
+
+from deepchem.feat.bio_seq_featurizer import SAMFeaturizer
+from deepchem.feat.bio_seq_featurizer import BAMFeaturizer
```
Comment (shreyasvinaya) [inline deepchem/feat/__init__.py:90]: need to add this in a try catch, might break imports if pysam is not installed

**Pull request closed without merging**